### PR TITLE
adds episode discussing how to enable reproducible use of containers

### PR DESCRIPTION
**Pull request by aturner-epcc**
The current version of the lesson has lots of excellent material on using Docker but when we initially ran a workshop using the lesson we thought that there was a gap where we needed to cover how you can use Docker containers to enable reproducibility. This episode is our attempt at producing material to cover that gap.

### CODE REVIEW

**Comment by ChristinaLK** [comment]
This is lovely! thanks @aturner-epcc. 

I really like how this puts the pieces together at the end - as you say in this section, we've really covered the mechanics in this lesson but this circles back to the goals discussed at the beginning, of making our lives easier by using containers. 

No major comments from me. @sstevens2 or @dme26 does one of you want to also review + merge?

**Review by ChristinaLK — APPROVED** [review]

File: _episodes/08-reproducibility.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+## Reproducibility
+
+By *reproducibility* here we mean the ability of someone else (or your future self) being able to reproduce what you did computationally at a particular time (be this in research, analysis or something else) as closely as possible even if you do not have access to exactly the same hardware resources that you had when you did the original work.
+
+Hopefully it is reasonably self-evident why containers are an attractive technology to help with reproducibility:
+
+  - The same computational work can be run across multiple different technologies seamlessly (e.g. Windows, macOS, Linux).
+  - You can save the exact process that you used for your computational work (rather than relying on potentially incomplete notes).
+  - You can save the exact versions of software and their dependencies in the image.
+  - You can access legacy versions of software and underlying dependencies which may not be generally available any more.
```
Comment (sstevens2) [inline _episodes/08-reproducibility.md:35]: Maybe we should add another reason below this line about also keeping a copy of the data with the scripts and analysis in an image?

File: _episodes/08-reproducibility.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+{: .bash}
+
+Once you have your archive, you can [deposit it on Zenodo](https://zenodo.org/deposit/) and this will:
+
+   - Create a long-term archive snapshot of your Docker image which people (including your future self) can download and reuse or reproduce your work.
+   - Create a persistent DOI (*Digital Object Identifier*) that you can cite in any publications or outputs to enable reproducibility and recognition of your work.
+
+In addition to the archive file itself, the deposit process will ask you to provide some basic metadata to classify the image and the associated work.
+
+Note that Zenodo is not the only option for archiving and generating persistent DOIs for images. There are other services out there - for example, some organizations may provide their own, equivalent, service.
```
Comment (sstevens2) [inline _episodes/08-reproducibility.md:73]: I love the addition of this section.  Should we write it for them to actually do this process with an image created earlier in the lesson?

File: _episodes/08-reproducibility.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  - Create many container images each running a single command (or step) of the workflow and use them in
+    sequence
+
+Of course, many real applications will sit somewhere between these two extremes.
+
+> ## Positives and negatives
+> What are the advantages and disadvantages of the two approaches to container granularity for research
+> workflows described above? Think about this
+> and write a few bullet points for advantages and disadvantages for each approach in the course Etherpad.
+{: .challenge}
```
Comment (sstevens2) [inline _episodes/08-reproducibility.md:98]: Would be good to have an answer for this challenge, even if it is exhaustive of all the possible answers learners might come up with.

File: _episodes/08-reproducibility.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+### Container Orchestration
+
+Although you can certainly manage research workflows that use multiple containers manually, there are a number of 
+container orchestration tools that you may find useful when managing workflows that use multiple containers. 
+Introducing the use of these orchestration tools is beyond the scope of this lesson so we briefly describe 
+a few options and point to useful resources on using these tools to allow you to explore them yourself. We
+cover:
```
Comment (sstevens2) [inline _episodes/08-reproducibility.md:106]:
```suggestion
Although you can certainly manage research workflows that use multiple containers manually, there are a number of 
container orchestration tools that you may find useful when managing workflows that use multiple containers. 
We won't go in depth on using these tools in this lesson but instead briefly describe 
a few options and point to useful resources on using these tools to allow you to explore them yourself. :
```

**Review by sstevens2 — CHANGES_REQUESTED** [review]
I think this episode is a great addition to the lesson.  Please take a look at my suggestions and then I'm happy to merge it in.

**Comment by aturner-epcc** [comment]
@sstevens2 Sorry for the delay. Thanks for the great comments - I think I have addressed them all in the latest commit.

File: _episodes/08-reproducibility.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+> ## Work in progress...
+> Note that reproducibility aspects of software and containers are an active area of research, discussion and development so are subject to many changes. We will present some ideas and approaches here but best practices will likely evolve in the near future.
+{: .callout}
+
+## Reproducibility
+
+By *reproducibility* here we mean the ability of someone else (or your future self) being able to reproduce
+what you did computationally at a particular time (be this in research, analysis or something else)
+as closely as possible even if you do not have access to exactly the same hardware resources #
```
Comment (sstevens2) [inline _episodes/08-reproducibility.md:32]: The sentence is written as someone else (except in parentheses) so I think it should be they here. 
```suggestion
as closely as possible even if they do not have access to exactly the same hardware resources #
```

**Comment by sstevens2** [comment]
Looks great! Made one small change and will go ahead and merge.

**Pull request merged**